packages/aws-client/src/S3.ts

```diff
@@ -779,39 +706,148 @@ export const uploadS3FileStream = (
       Body: fileStream,
     },
   });
+/**
+ * List of outputs from ListS3Objects functions (including V2)
+ * Guarantees that Key exists
+ */
+export interface ListS3ObjectsOutput extends _Object {
+  Key: string
+}
 
 /**
  * List the objects in an S3 bucket
- *
- * @param {string} bucket - The name of the bucket
- * @param {string} prefix - Only objects with keys starting with this prefix
- *   will be included (useful for searching folders in buckets, e.g., '/PDR')
- * @param {boolean} skipFolders - If true don't return objects that are folders
- *   (defaults to true)
- * @returns {Promise} A promise that resolves to the list of objects. Each S3
- *   object is represented as a JS object with the following attributes: `Key`,
- * `ETag`, `LastModified`, `Owner`, `Size`, `StorageClass`.
  */
 export const listS3Objects = async (
   bucket: string,
   prefix?: string,
   skipFolders: boolean = true
-) => {
+): Promise<Array<ListS3ObjectsOutput>> => {
   log.info(`Listing objects in s3://${bucket}`);
   const params: ListObjectsRequest = {
     Bucket: bucket,
   };
   if (prefix) params.Prefix = prefix;
 
   const data = await s3().listObjects(params);
-  let contents = data.Contents || [];
+  if (!data.Contents) {
+    return [];
+  }
+  let contents = data.Contents.filter((obj) => obj.Key !== undefined) as Array<ListS3ObjectsOutput>;
   if (skipFolders) {
     // Filter out any references to folders
-    contents = contents.filter((obj) => obj.Key !== undefined && !obj.Key.endsWith('/'));
+    contents = contents.filter((obj) => !obj.Key.endsWith('/'));
   }
   return contents;
 };
 
+/**
+ * Fetch complete list of S3 objects
+ *
+ * listObjectsV2 is limited to 1,000 results per call.  This function continues
+ * listing objects until there are no more to be fetched.
+ *
+ * The passed params must be compatible with the listObjectsV2 call.
+ *
+ * https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#listObjectsV2-property
+ *
+ * @param {Object} params - params for the s3.listObjectsV2 call
+ * @returns {Promise<Array>} resolves to an array of objects corresponding to
+ *   the Contents property of the listObjectsV2 response
+ *
+ * @static
+ */
+export const listS3ObjectsV2 = async (
+  params: ListObjectsV2Request
+): Promise<Array<ListS3ObjectsOutput>> => {
```

Comment:
The return type was Promise<ListObjectsV2Output['Contents']>.
Why do we need to filter on `obj.Key`? Every object has a key.

Reply:
while every object *should* have a key, that object is define with Key as optional. so this layer guarantees it and returns an object type which likewise guarantees it